### DISCUSSION TITLE
simple chaining of Keycloak

### DISCUSSION THREAD

**Discussion by mysterypersonx** [keycloak/keycloak#23599] [Q&A]
I'm trying to do a simple chaining of Keycloak.

I tried first with two realms. Now trying with two instances of keycloak on the same server, but different ports.

Let's call them: inner-keycloak and outer-keycloak.

The inner-keycloak instance is configured with User Federation, where LDAP servers have been configured. 
From this a SAML Token is generated. SAML client has been defined: this SAML Client is referring to Identity Provider defined in the outer-keycloak instance. Using the SP metadata from that Identity Provider definition in the outer-keycloak instance.

So, in the outer-keycloak instance, a SAML Identity Provider has been defined, using the Identity Provider metadata from the realm (containing corresponding client definition) in the inner-keycloak instance.
(Also Azure Active Directory is defined in a second Identity Provider in the outer-keycloak instance - this IDP works!)
An application has been defined as SAML client in this outer-keycloak instance. Users are to be authenticated by either Azure AD (doesn't contain all users) or the LDAP Server  (contains the missing users). 

Using a SAML tracer, we can see that the SAML Token gets correctly created from the LDAP server, in the inner-keycloak instance. But it doesn't get transferred to the IDP of the outer-keycloak instance.

Questions:
1. Can this 'chaining' work in Keycloak?
2. We could also use openid instead of SAML, between client in inner-keycloak instance and IDP in outer-keycloak instance. If that would solve the 'connection' issue between them.
3. I was hoping to accomplish this with two realms in one Keycloak instance, instead of two instances of Keycloak. Could it be done with realms, also?
4. Do I need to use the 'Token-Exchange' preview feature to accomplish the intended chaining? I have the feature enabled.
5. Any pointers/suggestions that I need to pay attention to? A setting that needs changing from the default?
6. Can you only accomplish this by creating a custom authentication flow? In the outer-keycloak instance only? Or also in the inner-keycloak instance? Have tried configuration to make it use the default 'first broker login' and 'browser' authentication flows - also didn't help. 

Any help/suggestions on this issue is very welcome.

**Comment by justin-tay** [comment]
It sounds like you just want to configure [Identity Brokers](https://www.keycloak.org/docs/latest/server_admin/#_identity_broker) so it's not clear what isn't working.

For instance there are Keycloak tests that configure a consumer realm and a provider realm for integration testing for the brokering.

There's no relation here with using token exchange.

**Reply by mysterypersonx** [reply]
It is Identity Brokering indeed. Thanks for the link.
And there are 2 Identity Providers being brokered. 
One is in the same realm, same instance, to Azure AD. The brokering works fine for this one. Client application gets the SAML Token from Azure AD via the IDP configuration for it. 
The other one (IDP configuration) is to another Keycloak realm and instance acting as the Identity Provider. There it fails to pass the authentication Token from that other Keycloak realm and instance. 

Looking at the documentation it fails on step 8 - I can see the SAML Token in the tracer from step 7 still.

**8. Keycloak checks if the response from the identity provider is valid. If valid, Keycloak imports and creates a user if the user does not already exist. Keycloak may ask the identity provider for further user information if the token does not contain that information. This behavior is identity federation. If the user already exists, Keycloak may ask the user to link the identity returned from the identity provider with the existing account. This behavior is account linking. With Keycloak, you can configure Account linking and specify it in the [First Login Flow](https://www.keycloak.org/docs/latest/server_admin/#_identity_broker_first_login). At this step, Keycloak authenticates the user and issues its token to access the requested resource in the service provider.**
 
How can I get Keycloak to just pass the Identification Token from one realm to another. Or from one Keycloak instance to another. Using SAML or OpenID. Trusting the Token from the other realm, or from the other instance?

**Reply by justin-tay** [reply]
I don't understand what is not working and what  you mean by pass the token. What is the error message? What token? As in the SAML assertion / OpenID ID Token? After Keycloak processes the brokered Identity Provider's response it will then issue its own SAML assertion / OpenID Token based on what [mappers](https://www.keycloak.org/docs/latest/server_admin/#_protocol-mappers) you have configured.

**Comment by mysterypersonx** [comment]
Error message I am getting now: 

![errmsg_kc](https://github.com/keycloak/keycloak/assets/146438135/29da693c-391c-4eb7-94d8-44f189a6783e)

I checked the mappers again. They are correct I believe. The application only needs: email, firstname and lastname. I have configured it as basic. 

This is the SAML Token generated by the "inner" Keycloak instance (port 9443), just before the error occurs, trying to send it back to the "outer" Keycloak instance (port 8443). 

![samltoken_kc](https://github.com/keycloak/keycloak/assets/146438135/4efd09b5-ef4b-4ecd-a227-896ae5778c4b)

In the "outer" Keycloak instance the mapping is defined as: 

![mapper_outerkc](https://github.com/keycloak/keycloak/assets/146438135/a5a4f6a3-02ac-4af7-b3e1-92187cf8577b)

**Reply by mysterypersonx** [reply]
Or is the SAML Token getting back to the IDP in the 'outer' Keycloak instance? But it is rejected or for some other reason not accepted?

**Reply by justin-tay** [reply]
The assertion is being rejected. You will need to check the logs to narrow it down.

https://github.com/keycloak/keycloak/blob/9b719006b07cb64231693133edb013d7ae1a2c42/services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java#L411-L598

**Reply by mysterypersonx** [reply]
that helped me a bit further. 
So the error was "validation failed". Looking at the section in the code, it is related to assertion signatures. I changed the settings of the client definition in the 'inner' Keycloak instance to have the assertion signed. 

![image](https://github.com/keycloak/keycloak/assets/146438135/73430cd9-bf16-4728-a842-0c4b019117ac)

and the IDP definition "idptest" in the 'outer' Keycloak instance to this: 

![image](https://github.com/keycloak/keycloak/assets/146438135/2915ec00-6ed7-47d5-8903-fd2f55a7541a)

The SAML tracer doesn't show an error anymore:

![image](https://github.com/keycloak/keycloak/assets/146438135/11516a8d-9f4d-458c-ae2e-de4bb25ee89b)

But at the end the 'outer' Keycloak instance redirects the users to this page. Asking email/username and password a second time. 

![image](https://github.com/keycloak/keycloak/assets/146438135/0197f1b1-3677-417f-95ea-70169a1af07e)

and if we give email and password again, it doesn't recognize it and gives error messages:

![image](https://github.com/keycloak/keycloak/assets/146438135/668636a4-d4f9-48ee-a6b4-dafca83d1211)

In the Keycloak log of the 'outer' Keycloak instance, we get these messages related to this: 

![image](https://github.com/keycloak/keycloak/assets/146438135/9b048bb4-6e6b-4020-a2cd-932f615153ca)

where the clientID and redirect_uri are the from the application defined as client in the 'outer' Keycloak instance. 

So, it seems to have gotten closer!

How can we avoid this second screen for entering username/password again in Keycloak? Is this related to the authentication flows? 

The authentication flows are defined as this in the 'inner' Keycloak instance for the client:

![image](https://github.com/keycloak/keycloak/assets/146438135/4eed6c2e-254d-45f1-8642-9368b652706b)

And the authentication flow is defined as this for the 'outer' Keycloak instance: 

![image](https://github.com/keycloak/keycloak/assets/146438135/365cd600-bcc9-4c26-966c-544db5d7e205)

**Reply by justin-tay** [reply]
Keycloak has detected that the user as defined in idptest already exists in azureadplus and so is asking to do account linkage and prompting for the credentials for logging in to azureadplus to do so. The linkage will need to be done once. Typically it will identify that a user with the same email or username is a unique user.

**Reply by mysterypersonx** [reply]
With the realm AZUREADPLUS selected, a search for the user that I used to test, gives nothing.
Can it be something else?

Also, do you need to create the user in the realm? Will that then happen to each user that authenticates via Keycloak? And if it is necessary, can you overwrite/refresh existing user in the Keycloak realm automatically, each time the user is authenticated? To avoid this 'screen' and question to link account to account in i

**Reply by justin-tay** [reply]
That would be rather unusual. This is following the default `first broker login` authentication flow. If the user doesn't exist it will create the user, if the user exists it then it will attempt to do the linkage.

You might want to read the documentation for [First login flow](https://www.keycloak.org/docs/latest/server_admin/#_identity_broker_first_login) and see if any of the options meets your needs. For instance [Disabling automatic user creation](https://www.keycloak.org/docs/latest/server_admin/#_disabling_automatic_user_creation).

![firstbroker](https://github.com/keycloak/keycloak/assets/49700559/2a8ed17d-c532-4956-8d9d-c5d26f129459)

**Comment by skn33** [comment]
dont know if it has been solved, anyway I have been facing the same issue, the pb is that I have updated the alias of the IDP, when u change the alias, the redirect URI is changed automatically, so need to update the URL in the SAML IDP
